### PR TITLE
feat: Support versions on random positions

### PR DESCRIPTION
**Pull request by jaysonsantos**
First thanks for this quite good project!

## Description
This is a rather confusing logic but I decided to send this PR as a way to start the conversation.
I am trying to use commitizen to bump rust projects and its default way works fine when bumping the version inside a `Cargo.toml` as usually, your own package's version is on the top. But if you have a lock file, rust also bumps the version there. This PR was done in a way to bump the version only after something is matched, in the example, I used the project name so the versions_file could have something like `Cargo.lock:version:project-name` where it only bumps version if the project name was found before it.

## Checklist

- [x] Add test cases to all the changes you introduce
- [x] Run `./script/format` and `./script/test` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [ ] Update the documentation for the changes

## Expected behavior
Bumping versions on `Cargo.lock` or any other file, should bump only the intended line.

## Steps to Test This Pull Request
Try to bump any Cargo.lock file or any file that can have your own version, randomly placed in it.

### CODE REVIEW

**Comment by Lee-W** [comment]
Hi @jaysonsantos, thanks for this PR. This one looks like a rust-specific solution to me. I'm thinking of using regex like `name = "there-i-fixed-it"\nversion = "\d\.\d\.\d\."`'. But currently, we're paring it by line. We might need to re-implement this method

**Comment by jaysonsantos** [comment]
Hey there @Lee-W that was my first try and then I found out it was line based.
What if I change it to regex replace on the whole file with limit of one?

**Comment by Woile** [comment]
Hey thanks for the contribution, this looks like a nice feature to have.

I think a re-implementation of the line based regex would be really cool , but we should not forget the no regex case.
Looking forward to see it 💪🏻

**Comment by Lee-W** [comment]
I don't have a clear idea of how this should be implemented. Maybe we could discuss more on this thread before we implement the new version 🙂

**Comment by jaysonsantos** [comment]
One of the ways I see this working is:
- Read the whole file contents into a variable
- Search with a regex so you get the start-end of the match
- Seek to the end of the match, get everything from there until the next \n or eof
- Replace only between the end of the match until \n or eof

**Comment by jaysonsantos** [comment]
Hey there folks, check the latest commit I sent to see the implementation https://github.com/commitizen-tools/commitizen/pull/361/commits/4ee04dcf6496ba061067bd58cfe651b342326068

**Comment by jaysonsantos** [comment]
Sorry, I forgot that bumpversion uses a bunch of python versions and I used the walrus operator

**Comment by jaysonsantos** [comment]
Ok, no walrus operator now

**Comment by Lee-W** [comment]
@jaysonsantos Thanks for your immediate update 🎉 The behavior is slightly changed (update all v.s. update the last one). I'm not sure whether we should consider it as a breaking change. We might need to describe it in our documentation clearly. Also, I'm thinking of displaying a warning if there're multiple regex matches.

**Comment by jaysonsantos** [comment]
> The behavior is slightly changed (update all v.s. update the last one).

I didn't get I think it is the same logic, no?
The old one would loop through all lines and replace exact matches of the old version to the new one. This https://github.com/commitizen-tools/commitizen/pull/361/files#diff-ed9bead31489109a8441fbfb00d2ddec00198489b0702bab164825c2d3bd9b4fR191 will be even more strict as it will only update if the exact version is surrounded by boundaries, so v1.2.3 wouldn't be changed.
I **THINK** it does not break compatibility, if you can think of a test, can you send me and I add a case for it?

File: commitizen/bump.py [hunk 1]
```diff
@@ -150,26 +150,25 @@ def update_version_in_files(
...
-                        continue
-
-                # Replace the target string
-                if current_version in line:
-                    current_version_found = True
-                    file_content.append(line.replace(current_version, new_version))
-                else:
-                    file_content.append(line)
+
+        version_file = open(filepath, "r").read()
```
Comment (Lee-W) [inline commitizen/bump.py:157]: Should we use `with open` here?

File: commitizen/bump.py [hunk 2]
```diff
@@ -150,26 +150,25 @@ def update_version_in_files(
...
+            right = version_file[match.end() :]
+            line_break = _get_line_break_position(right)
+            middle = right[:line_break]
+            current_version_found = current_version in middle
+            right = right[line_break:]
+            version_file = left + middle.replace(current_version, new_version) + right
+
+        if not regex:
+            current_version_regex = _version_to_regex(current_version)
+            current_version_found = current_version_regex.search(version_file) and True
```
Comment (Lee-W) [inline commitizen/bump.py:170]: What is this `and True` for?

**Comment by Lee-W** [comment]
> > The behavior is slightly changed (update all v.s. update the last one).
> 
> I didn't get I think it is the same logic, no?
> The old one would loop through all lines and replace exact matches of the old version to the new one. This https://github.com/commitizen-tools/commitizen/pull/361/files#diff-ed9bead31489109a8441fbfb00d2ddec00198489b0702bab164825c2d3bd9b4fR191 will be even more strict as it will only update if the exact version is surrounded by boundaries, so v1.2.3 wouldn't be changed.
> I **THINK** it does not break compatibility, if you can think of a test, can you send me and I add a case for it?

I'm actually ok with it. Just raise this issue as I was not yet familiar enough with the changes.

**Comment by jaysonsantos** [comment]
@Lee-W you can take a look again

**Review by Lee-W — APPROVED** [review]
Looks good to me 😃

**Pull request merged**